### PR TITLE
[NT-776] Add Optimizely Tracking Events

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Adds `Project Page Rewards CTA Tapped` and `Pledge Screen Viewed` to be tracked via Optimizely.

# 🤔 Why

In order to track the success of our _See rewards_ experiment.

# 🛠 How

- Adds these two events and their respective attributes and event tags.
- Added `track(eventKey:userId:attributes:eventTags:)` to `OptimizelyClientType` in order to test using our `MockOptimizelyClient`.

# ✅ Acceptance criteria

- [ ] Confirm that the attributes are correct.
- [ ] Confirm that these events are being tracked to Optimizely.

### CODE REVIEW

File: Kickstarter-iOS/Library/CookieRefTagFunctions.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import Foundation
+import KsApi
+
+private let cookieSeparator = "?"
+private let escapedCookieSeparator = "%3F"
+
+// Extracts the ref tag stored in cookies for a particular project. Returns `nil` if no such cookie has
+// been previously set.
+public func cookieRefTagFor(project: Project) -> RefTag? {
```
Comment (justinswart) [inline Kickstarter-iOS/Library/CookieRefTagFunctions.swift]: Moved these functions to a shared place now that we need to access `cookieRefTagFor` in two places. Now that they're shared we may want to add tests for them specifically but I figured we could tackle them separately as this doesn't reduce our test coverage currently.

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/OptimizelyClientType.swift [hunk 2]
```diff
@@ -32,3 +35,28 @@ extension OptimizelyClientType {
...
+    "location": user?.location?.country.lowercased(),
+    "os_version": AppEnvironment.current.device.systemVersion,
+    "logged_in": user != nil,
+    "chosen_currency": project.stats.currentCurrency,
+    "locale": AppEnvironment.current.locale.identifier
+  ]
+  .compact()
+
+  let eventTags: [String: Any] = [
+    "project_subcategory": project.category.name,
```
Comment (justinswart) [inline Library/OptimizelyClientType.swift]: This _might_ not be the property that we want for this. The spreadsheet mentions "What project category has the user pledged previously?" - I've asked for confirmation.

File: Library/ViewModels/ProjectPamphletViewModel.swift [hunk 3]
```diff
@@ -217,6 +217,25 @@ public final class ProjectPamphletViewModel: ProjectPamphletViewModelType, Proje
       .map(cookieFrom(refTag:project:))
       .skipNil()
       .observeValues { AppEnvironment.current.cookieStorage.setCookie($0) }
+
+    Signal.combineLatest(project, refTag)
+      .takeWhen(ctaButtonTappedWithType.filter { $0 == .pledge })
+      .filter { project, _ in project.personalization.backing == nil }
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/ProjectPamphletViewModel.swift]: Yeah probably redundant!

File: Library/ViewModels/PledgeCTAContainerViewViewModel.swift [hunk 4]
```diff
@@ -130,19 +130,18 @@ public final class PledgeCTAContainerViewViewModel: PledgeCTAContainerViewViewMo
...
-      userId: uuid,
-      isAdmin: AppEnvironment.current.currentUser?.isAdmin ?? false
-    )
-
   guard let projectBacking = backing, project.personalization.isBacking == .some(true) else {
     if currentUserIsCreator(of: project) {
       return PledgeStateCTAType.viewYourRewards
     }
 
+    let optimizelyVariant = AppEnvironment.current.optimizelyClient?
```
Comment (ifbarrera) [inline Library/ViewModels/PledgeCTAContainerViewViewModel.swift]: Moved the activation of the experiment to _after_ the `guard` that checks whether the project has already been backed or not - this prevents us from activating the experiment when we know it won't be shown.

File: Library/ViewModels/ProjectPamphletViewModel.swift [hunk 5]
```diff
@@ -217,6 +217,27 @@ public final class ProjectPamphletViewModel: ProjectPamphletViewModelType, Proje
       .map(cookieFrom(refTag:project:))
       .skipNil()
       .observeValues { AppEnvironment.current.cookieStorage.setCookie($0) }
+
+    let shouldTrackCTATappedEvent = ctaButtonTappedWithType
+      .filter { [.pledge, .seeTheRewards, .viewTheRewards].contains($0) }
```
Comment (ifbarrera) [inline Library/ViewModels/ProjectPamphletViewModel.swift]: We weren't actually triggering the event on the experimental variants. Added the here and added tests to make sure we're tracking taps to the button in all three cases.

Comment (justinswart) [reply Library/ViewModels/ProjectPamphletViewModel.swift]: Do you think we should do this or just filter on the project not being backed yet? That way we don't need to update this when variants change 🤔

Comment (ifbarrera) [reply Library/ViewModels/ProjectPamphletViewModel.swift]: I think we need to be explicit about when we track an "impression" for an experimental variant being acted on. If we don't know for sure that a variant has been seen/activated, then we shouldn't track an event in Optimizely.

**Pull request merged**